### PR TITLE
Feat/webpack vendors bundles

### PR DESCRIPTION
**Pull request by skjnldsv**
- Reduce compile time by a power of 4
- Move all bundles into `/dist`
- Easier to cleanup
- Easier to target old legacy js files
- No more `vue-` or `js/dist` hacks
- Bundle all shared modules into one big cached file

### CODE REVIEW

File: core/src/OCP/collaboration.js [hunk 1]
```diff
@@ -25,20 +25,21 @@ import escapeHTML from 'escape-html'
 
 /**
  * @typedef TypeDefinition
- * @method {callback} action This action is executed to let the user select a resource
+ * @function {callback} action This action is executed to let the user select a resource
  * @param {string} icon Contains the icon css class for the type
- * @constructor
+ * @function Object() { [native code] }
```
Comment (artonge) [inline core/src/OCP/collaboration.js:30]: Not sure about this change

File: lib/private/legacy/OC_Template.php [hunk 2]
```diff
@@ -106,19 +106,22 @@ public static function initTemplateEngine($renderAs) {
...
-				OC_Util::addScript('dist/files_client', null, true);
-				OC_Util::addScript('dist/files_fileinfo', null, true);
+				OC_Util::addScript('core', 'merged-template-prepend', true);
+				OC_Util::addScript('core', 'files_client', true);
+				OC_Util::addScript('core', 'files_fileinfo', true);
 			}
-			OC_Util::addScript('core', 'dist/main', true);
+
+			// include common nextcloud webpack bundle
+			OC_Util::addScript('nextcloud', 'common', true);
```
Comment (artonge) [inline lib/private/legacy/OC_Template.php:123]: Does this work with production builds? I remember that the file name was looking like `nextcloud-0.js`.

File: webpack.common.js [hunk 3]
```diff
@@ -1,47 +1,27 @@
 /* eslint-disable camelcase */
-const { merge } = require('webpack-merge')
 const { VueLoaderPlugin } = require('vue-loader')
 const path = require('path')
-
+const CircularDependencyPlugin = require('circular-dependency-plugin')
```
Comment (artonge) [inline webpack.common.js:4]: Should we remove that plugin ?

File: lib/private/legacy/OC_Template.php (hunk 2 above)
Comment (skjnldsv) [reply lib/private/legacy/OC_Template.php:123]: yes, its name changed :)

File: .eslintrc.js [hunk 4]
```diff
@@ -16,5 +16,9 @@ module.exports = {
 		// TODO: make sure we fix this as this is bad vue coding style.
 		// Use proper sync modifier
 		'vue/no-mutating-props': 'warn',
+		'vue/custom-event-name-casing': ['error', 'kebab-case', {
+			// allows custom xxxx:xxx events formats
+			ignores: ['/^[a-z]+(?:-[a-z]+)*:[a-z]+(?:-[a-z]+)*$/u'],
```
Comment (Pytal) [inline .eslintrc.js:21]: This may be shared https://github.com/nextcloud/eslint-config/blob/master/index.js#L98 I think?

Comment (skjnldsv) [reply .eslintrc.js:21]: Yep!

Comment (Pytal) [reply .eslintrc.js:21]: https://github.com/nextcloud/eslint-config/pull/249

Comment (skjnldsv) [reply .eslintrc.js:21]: You rock! :hugs:

**Comment by skjnldsv** [comment]
Still an issue with the mapfile of the big chunk:
```
Error:
Run git status
HEAD detached at pull/30020/merge
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   dist/core-common.js.map
```

**Comment by skjnldsv** [comment]
All the issues seems to come from the `vue-hot-reload-api`
image

**Comment by skjnldsv** [comment]
Rebased

**Comment by artonge** [comment]
Fix webpack version because: webpack/webpack#14628

**Comment by artonge** [comment]
Rebased and compiled

**Comment by artonge** [comment]
Starting to see the end of it:

- [ ] 13 front end tests failing, down from 50. I will solve that.
- [ ] 1 PHPUnit test fails since rebase. Probably because of the conjugated work of @mejo- and @skjnldsv . Do you think you can take a look?
    - https://github.com/nextcloud/server/pull/30385
    - https://github.com/nextcloud/server/runs/4704563953?check_suite_focus=true#step:7:352
- [ ] `common-main.js` is included twice in the HTML template. @skjnldsv You probably know more than me about that, and it's probably related to the above error. If you think that you can solve this in a few minutes, can you take a look at it?

Screenshot from 2022-01-05 19-51-05

**Comment by artonge** [comment]
Fixed undescore version because of: #27279

**Comment by artonge** [comment]
Detect app's script higher up as some scripts were shadowed:

Add in `lib/private/Template/JSResourceLocator.php`
```php
|| $this->appendIfExist($this->serverroot, 'apps/'.$script.'.js')
```

**Comment by skjnldsv** [comment]
One last failure in phpunit:

```
There was 1 failure:
520 |  
521 | 1) Test\UtilTest::testAddScript
522 | Failed asserting that 14 matches expected 0.
523 |  
524 | /drone/src/tests/lib/UtilTest.php:259
```

File: package.json [hunk 5]
```diff
@@ -95,28 +98,32 @@
...
+    "@nextcloud/eslint-config": "^6.1.2",
     "@nextcloud/stylelint-config": "^2.0.1",
     "@testing-library/jest-dom": "^5.16.1",
     "@testing-library/user-event": "^13.5.0",
     "@testing-library/vue": "^5.8.2",
+    "@vue/test-utils": "^1.3.0",
     "babel-loader": "^8.2.3",
     "babel-loader-exclude-node-modules-except": "^1.2.1",
     "css-loader": "^5.2.7",
     "eslint-loader": "^4.0.2",
```
Comment (skjnldsv) [inline package.json]: Can be dropped if we use `eslint-webpack-plugin`
Can be a followup :wink:

File: lib/private/legacy/template/functions.php [hunk 6]
```diff
@@ -111,17 +114,21 @@ function print_unescaped($string) {
 
 /**
  * Shortcut for adding scripts to a page
+ * All scripts are forced to be loaded after core since
+ * they are coming from a template registration.
+ * Please consider moving them into the relevant controller
```
Comment (skjnldsv) [inline lib/private/legacy/template/functions.php]: We should deprecate

**Comment by skjnldsv** [comment]
My PR, but it's a go from me :rocket:

File: package.json [hunk 7]
```diff
@@ -68,13 +69,15 @@
     "moment-timezone": "^0.5.34",
     "nextcloud-vue-collections": "^0.9.0",
     "p-limit": "^4.0.0",
-    "p-queue": "^7.1.0",
+    "p-queue": "^6.6.2",
```
Comment (MichaIng) [inline package.json]: Was this downgrade intended/required?

Comment (skjnldsv) [reply package.json]: Yes, I think it's because there is an export issue with latst p-queue 
https://github.com/sindresorhus/p-queue/issues/145

**Comment by artonge** [comment]
> EDIT: weird, I can't reproduce locally

It is reproducible when executing all the tests as it is a state error. Seen with jonas who came up with a fix. Should be good now :)

**Comment by skjnldsv** [comment]
Yes, passes now, LGTM!

**Comment by mejo-** [comment]
> Yes, passes now, LGTM!
> 
> EDIT NOT grin :
> 
> ```
> 33) Test\UtilTest::testFilenameValidation with data set #21 ('double"quote', true)
> Error: Access to undeclared static property: OC_Util::$scriptDeps
> ```

Sorry for the hickup :( I can look into it on monday. Today I'm not working unfortunately.

File: lib/public/Util.php [hunk 8]
```diff
@@ -597,4 +597,13 @@ public static function shortenMultibyteString(string $subject, int $dataLength,
 		}
 		return $temp;
 	}
+
+	/**
+	 * Reset scripts and scriptDeps to empty arrays.
+	 * Only used in tests.
+	 */
```
Comment (mejo-) [inline lib/public/Util.php:604]: You probably have to add `@since 24.0.0` here.

Comment (artonge) [reply lib/public/Util.php:604]: Thanks for the tip :)

File: lib/private/legacy/template/functions.php (hunk 6 above)
Comment (artonge) [reply lib/private/legacy/template/functions.php:119]: https://github.com/nextcloud/server/pull/30020/commits/7b5ecf4786ab42512a424d6dc979caabae265e76

File: package.json (hunk 7 above)
Comment (artonge) [reply package.json]: Fixed it, eslint: dumb, webpack: smart

**Comment by artonge** [comment]
:warning: CI failures might not be unrelated this time. Never seen them. Need to check.

**Comment by skjnldsv** [comment]
Rebased

**Comment by skjnldsv** [comment]
```
Scenario: delete a user                                                      # /drone/src/tests/acceptance/features/users.feature:25
130 | Given I act as Jane                                                        # ActorContext::iActAs()
131 | And I am logged in as the admin                                            # LoginPageContext::iAmLoggedInAsTheAdmin()
132 | And I open the User settings                                               # SettingsMenuContext::iOpenTheUserSettings()
133 | And I see that the list of users contains the user user0                   # UsersSettingsContext::iSeeThatTheListOfUsersContainsTheUser()
134 | And I open the actions menu for the user user0                             # UsersSettingsContext::iOpenTheActionsMenuOf()
135 | And I see that the "Delete user" action in the user0 actions menu is shown # UsersSettingsContext::iSeeTheAction()
136 | Failed asserting that false is true.
137 | When I click the "Delete user" action in the user0 actions menu            # UsersSettingsContext::iClickTheAction()
138 | And I click the "Delete user0's account" button of the confirmation dialog # DialogContext::iClickTheDialogButton()
139 | Then I see that the list of users does not contains the user user0         # UsersSettingsContext::iSeeThatTheListOfUsersDoesNotContainsTheUser()
140
```

Is valid, the menu does not open anymore.

File: lib/private/legacy/template/functions.php [hunk 9]
```diff
@@ -111,17 +114,22 @@ function print_unescaped($string) {
...
+ *
  * @param string $app the appname
  * @param string|string[] $file the filename,
  * if an array is given it will add all scripts
+ * @deprecated 23.1.0
```
Comment (skjnldsv) [inline lib/private/legacy/template/functions.php]: Why 23.1.0 ?
Should be 24 ?

File: lib/public/Util.php [hunk 10]
```diff
@@ -597,4 +597,14 @@ public static function shortenMultibyteString(string $subject, int $dataLength,
 		}
 		return $temp;
 	}
+
+	/**
+	 * Reset scripts and scriptDeps to empty arrays.
+	 * Only used in tests.
+	 * @since 24.0.0
+	 */
+	public static function clearScriptsState() {
```
Comment (skjnldsv) [inline lib/public/Util.php]: I would prefer a different way of clearing private static methods.
Phpunit have a way for that iirc. We can do it in a followup :pray:

File: lib/private/legacy/template/functions.php (hunk 9 above)
Comment (artonge) [reply lib/private/legacy/template/functions.php]: True, my bad

Comment (skjnldsv) [reply lib/private/legacy/template/functions.php]: Please fix on followup so we don't yet rebase this monstrous PR :see_no_evil:

**Review by artonge — APPROVED** [review]

**Review by CarlSchwan — APPROVED** [review]

File: package.json (hunk 5 above)
Comment (artonge) [reply package.json:110]: Do you mean `@testing-library/vue` instead of `eslint-webpack-plugin` ?

Comment (artonge) [reply package.json:110]: Oh, the green line is not the one you reference...

File: lib/private/legacy/template/functions.php (hunk 9 above)
Comment (blizzz) [reply lib/private/legacy/template/functions.php]: Also, please add a pointer to its replacement.

**Pull request merged**